### PR TITLE
Amend the wrong data index of ScaleBIBitmap method

### PR DESCRIPTION
**Pull request by hjm1fb**
Fix the bug of having Aliasing when scaling down an image using Bilinear Interpolation at certain rates like 0.6 by doing 
1. Amend the wrong offset of rgbTopRight point.
2. Do more accurate computing on sampling point.

### CODE REVIEW

**Comment by AndroidDeveloperLB** [comment]
Can you please show the problem ?
Maybe a sample project? Screenshots of before and after, together with the original image?

File: JniBitmapOperationsLibrary/jni/JniBitmapOperationsLibrary.cpp [hunk 1]
```diff
@@ -422,7 +426,7 @@ JNIEXPORT void JNICALL Java_com_jni_bitmap_1operations_JniBitmapHolder_jniScaleB
 			&rgbTopLeft);
 		//rgbTopRight=startingImageData[xTopLeft+1][yTopLeft];
 		convertIntToArgb(
-			previousData[((yTopLeft + 1) * oldWidth) + xTopLeft],
```
Comment (hjm1fb) [inline JniBitmapOperationsLibrary/jni/JniBitmapOperationsLibrary.cpp]: This is obvious that the offset of "top-right" position should be  yTopLeft*width + xTopLeft + 1

**Comment by hjm1fb** [comment]
Sorry, I can't do that demo right now for some reason. But I added a comment just now. That is, "This is obvious that the offset of "top-right" position should be yTopLeft*width + xTopLeft + 1".
At least that modification is safe. 
And thanks for your great code~

File: JniBitmapOperationsLibrary/jni/JniBitmapOperationsLibrary.cpp [hunk 2]
```diff
@@ -422,7 +426,7 @@ JNIEXPORT void JNICALL Java_com_jni_bitmap_1operations_JniBitmapHolder_jniScaleB
...
+			previousData[(yTopLeft* oldWidth) + xTopLeft + 1],
 			&rgbTopRight);
```
Comment (hjm1fb) [inline JniBitmapOperationsLibrary/jni/JniBitmapOperationsLibrary.cpp]: This is obvious that the offset of "top-right" position should be yTopLeft*width + xTopLeft + 1

**Comment by AndroidDeveloperLB** [comment]
I don't think changing from `xTopLeft = oldWidth - 2; ` to `xTopLeft--;` is better. I made sure that it just stays on the right boundaries.
Why did you remove the code of setting `xt` on this condition? 

As for calculation of rgbTopRight, I wrote this explanation:
		//rgbTopRight=startingImageData[xTopLeft+1][yTopLeft];

So it's supposed to add to X just 1. Not an entire row. You wrote this:
```
previousData[((yTopLeft + 1) * oldWidth) + xTopLeft] ...
```
This would mean that you go down a row. And what you wrote is actually the same as I did for a different case :
```
//rgbBottomLeft=startingImageData[xTopLeft][yTopLeft+1];
```
https://github.com/AndroidDeveloperLB/AndroidJniBitmapOperations/blob/4e9f3a5e843c098efe031ea97729d6bec9bc01f6/JniBitmapOperationsLibrary/jni/JniBitmapOperationsLibrary.cpp#L404

Sorry, but it sounds incorrect to me at this point. You should show me the issue you are seeing that caused you to write it, and explain better why you think what you wrote solves it.

**Comment by hjm1fb** [comment]
> I don't think changing from `xTopLeft = oldWidth - 2; ` to `xTopLeft--;` is better. I made sure that it just stays on the right boundaries.
> Why did you remove the code of setting `xt` on this condition?
> 
> As for calculation of rgbTopRight, I wrote this explanation:
> //rgbTopRight=startingImageData[xTopLeft+1][yTopLeft];
> 
> So it's supposed to add to X just 1. Not an entire row. You wrote this:
> 
> ```
> previousData[((yTopLeft + 1) * oldWidth) + xTopLeft] ...
> ```
> 
> This would mean that you go down a row. And what you wrote is actually the same as I did for a different case :
> 
> ```
> //rgbBottomLeft=startingImageData[xTopLeft][yTopLeft+1];
> ```
> 
> https://github.com/AndroidDeveloperLB/AndroidJniBitmapOperations/blob/4e9f3a5e843c098efe031ea97729d6bec9bc01f6/JniBitmapOperationsLibrary/jni/JniBitmapOperationsLibrary.cpp#L404
> 
> Sorry, but it sounds incorrect to me at this point. You should show me the issue you are seeing that caused you to write it, and explain better why you think what you wrote solves it.

I think we have some misunderstanding here.
The previous line is 
`//rgbTopRight=startingImageData[xTopLeft+1][yTopLeft];
		convertIntToArgb(
			previousData[((yTopLeft + 1) * oldWidth) + xTopLeft],
			&rgbTopRight);`
which go down a row.
And I made it to 
`//rgbTopRight=startingImageData[xTopLeft+1][yTopLeft];
		convertIntToArgb(
			previousData[(yTopLeft* oldWidth) + xTopLeft + 1],
			&rgbTopRight);`
which is same with your intention 
`startingImageData[xTopLeft+1][yTopLeft]`

**Comment by AndroidDeveloperLB** [comment]
Oh, so it's the opposite?

**Comment by hjm1fb** [comment]
indeed... you can check the original code here [JniBitmapOperationsLibrary.cpp](https://github.com/AndroidDeveloperLB/AndroidJniBitmapOperations/blob/master/JniBitmapOperationsLibrary/jni/JniBitmapOperationsLibrary.cpp) line 424.

**Comment by hjm1fb** [comment]
OK, I have pulled~

**Comment by AndroidDeveloperLB** [comment]
I guess the conversion from Java wasn't right, then.
I actually planned to convert more, from [Seam Carving resizing](http://www.faculty.idc.ac.il/arik/SCWeb/imret/index.html), but it was very inefficient in Java, that I wanted to optimize Java first, and then stopped because of how complex it is and how easy it is to make mistakes...

**Comment by hjm1fb** [comment]
It's OK. We all make bugs sometimes, I benefit from your code :)

**Comment by AndroidDeveloperLB** [comment]
I actually have no idea how to work with C/C++ on Android Studio anymore. Things seem to have changed a lot.
I hope it's not hard.
Did you have to use a lot of steps to handle the project?

**Comment by hjm1fb** [comment]
Nope, I just copy the C++ code in the method. Anything else like Java code, JNI interface, I managed myself. And as your code works fine now, I have no more workload on image scaling for now.

**Comment by hjm1fb** [comment]
> [Seam Carving resizing](http://www.faculty.idc.ac.il/arik/SCWeb/imret/index.html),

By the way, have you tried Seam Carving resizing on Android? Is it reasonable to apply that complex algorithm on Android? I am curious about different ways of resizing an image on Android (*￣︶￣)

**Comment by hjm1fb** [comment]
> I guess the conversion from Java wasn't right, then.
> I actually planned to convert more, from [Seam Carving resizing](http://www.faculty.idc.ac.il/arik/SCWeb/imret/index.html), but it was very inefficient in Java, that I wanted to optimize Java first, and then stopped because of how complex it is and how easy it is to make mistakes...

By the way, have you tried Seam Carving resizing on Android? Is it reasonable to apply that complex algorithm on Android? I am curious about different ways of resizing an image on Android (*￣︶￣)

**Comment by AndroidDeveloperLB** [comment]
Can you please make a video to show how to use JNI nowadays on Android Studio? Even Hello-world could be nice...

As for Seam Carving, it was back in the university. Was a group project and we didn't know how to split the work, and it was in a time that we didn't even know much about multi-threads. I was in charge of the 2 resizing techniques I've shared here, including some UI (using SWT), and someone else was in charge of Seam Carving. It worked, but it wasn't implemented efficiently at all. 
If you wish, I think this was the original code:

[CG1 original.zip](https://github.com/AndroidDeveloperLB/AndroidJniBitmapOperations/files/4345125/CG1.original.zip)

As for running on Android, I'm sure it's very possible as today's smartphones are very good. Plus it wasn't optimized well at all, not even for Java.

There are even apps out there that use this technique to remove object from the image, as you can see on [the video](https://www.youtube.com/watch?time_continue=8&v=6NcIJXTlugc) . I'm sure it's available on some image editing apps, too.

**Comment by hjm1fb** [comment]
> the video

OK, I will find an instruction video of using JNI in Android Studio for you. Or maybe an Android Studio Official JNI quick-start document.

**Comment by AndroidDeveloperLB** [comment]
The video doesn't show how it's done on new Android Studio versions:
https://youtu.be/Wua3yq3vVDk?t=270
I can't find "include C++ support" checkbox:
[2020-03-20_15-27-30.zip](https://github.com/AndroidDeveloperLB/AndroidJniBitmapOperations/files/4359781/2020-03-20_15-27-30.zip)

Can you please try on latest Android Studio version , and show a video?

**Comment by hjm1fb** [comment]
> The video doesn't show how it's done on new Android Studio versions:
> https://youtu.be/Wua3yq3vVDk?t=270
> I can't find "include C++ support" checkbox:
> [2020-03-20_15-27-30.zip](https://github.com/AndroidDeveloperLB/AndroidJniBitmapOperations/files/4359781/2020-03-20_15-27-30.zip)
> 
> Can you please try on latest Android Studio version , and show a video?

OK, I will make that video within one or two weeks. I also think instant message and instruction will be more efficient, how about friending me on one of these: Wechat, Skype, TeamView or Facebook?

**Comment by hjm1fb** [comment]
> The video doesn't show how it's done on new Android Studio versions:
> https://youtu.be/Wua3yq3vVDk?t=270
> I can't find "include C++ support" checkbox:
> [2020-03-20_15-27-30.zip](https://github.com/AndroidDeveloperLB/AndroidJniBitmapOperations/files/4359781/2020-03-20_15-27-30.zip)
> 
> Can you please try on latest Android Studio version , and show a video?
Sorry, I have never recorded instruction-video before. And the latest Android Studio has bad reputation of RAM consuming and dragging down the computer performance, as far as I know. So I come up two other options.
1. I can instruct you to do the C++ integration in Android Studio personally  through communication, by texting or voice call. As my English is OK both in Writing  and Speaking.
2. I can spend about two hours to refactor this project into CMake version one, so it can be directly built and run after imported into Android Studio of nowadays.

**Pull request closed without merging**